aten/src/ATen/native/attn.cpp

```diff
@@ -0,0 +1,86 @@
+#include <ATen/ATen.h>
+
+namespace at {
+namespace native {
+
+std::tuple<Tensor, Tensor> attn(
+    const Tensor& q,
+    const Tensor& k,
+    const Tensor& v) {
+
+  TORCH_CHECK(q.dim() == 2 && k.dim() == 2 && v.dim() == 2, "All of q,k,v need to be 2D tensor");
+  TORCH_CHECK(q.size(1) == k.size(1), "q and k must have must same size for 2nd dim");
+  TORCH_CHECK(k.size(0) == v.size(0), "k and v must have must same size for 1st dim");
+
+  Tensor x = at::matmul(q, k.t());
+  Tensor a = at::tanh(x);
+  Tensor o = at::matmul(a, v);
+
+  return std::make_tuple(o, a);
+}
+
+std::tuple<Tensor, Tensor, Tensor> attn_backward(
+    const Tensor& do_ext,
+    const Tensor& da_ext,
+    const Tensor& q,
+    const Tensor& k,
+    const Tensor& v,
+    const Tensor& a) {
+
+  Tensor dq, dk, dv;
+
+  // Alternatively, this check can be expressed in derivatives.yaml as
+  // - q, k, v: "grads[0].defined() || grads[1].defined() ? attn_backward(grads[0], grads[1], q, k, v, a) : std::tuple<Tensor, Tensor, Tensor>()"
+  if (!do_ext.defined() && !da_ext.defined()) {
+    return std::make_tuple(dq, dk, dv);
+  }
+
+  bool needs_dq = q.requires_grad();
```

Comment:
You should not do it this way. You should use the `grad_input_mask` variable available to you  in derivatives.yaml and pass it to this function. You can see derivatives.yaml for use of it.